File: kotlin-archive/jam-pvm/src/main/kotlin/io/forge/jam/pvm/program/JumpTable.kt
package io.forge.jam.pvm.program

import io.forge.jam.pvm.Abi

/**
 * Represents a table of jump addresses with fixed-size entries.
 *
 * @property blob The underlying byte array containing jump addresses
 * @property entrySize Size of each jump table entry in bytes
 */
@JvmInline
value class JumpTable(
    private val state: Pair<ByteArray, UInt>
) : Iterable<ProgramCounter> {
    private val blob: ByteArray get() = state.first
    private val entrySize: UInt get() = state.second

    companion object {
        /**
         * Creates a new JumpTable instance
         */
        fun create(blob: ByteArray, entrySize: UInt): JumpTable =
            JumpTable(Pair(blob, entrySize))
    }

    /**
     * Checks if the jump table is empty
     */
    fun isEmpty(): Boolean = len() == 0u

    /**
     * Returns the number of entries in the jump table
     */
    fun len(): UInt = if (entrySize == 0u) {
        0u
    } else {
        (blob.size.toUInt() / entrySize)
    }

    /**
     * Gets a program counter by memory address
     */
    fun getByAddress(address: UInt): ProgramCounter? {
        if (address and (Abi.VM_CODE_ADDRESS_ALIGNMENT - 1u) != 0u || address == 0u) {
            return null
        }
        return getByIndex(
            (address - Abi.VM_CODE_ADDRESS_ALIGNMENT) / Abi.VM_CODE_ADDRESS_ALIGNMENT
        )
    }

    /**
     * Gets a program counter by index
     */
    fun getByIndex(index: UInt): ProgramCounter? {
        if (entrySize == 0u) {
            return null
        }

        val start = index.toULong() * entrySize.toULong()
        if (start > Int.MAX_VALUE.toULong()) return null

        val end = start + entrySize.toULong()
        if (end > Int.MAX_VALUE.toULong()) return null

        val startInt = start.toInt()
        val endInt = end.toInt()

        if (startInt >= blob.size || endInt > blob.size) {
            return null
        }

        val slice = blob.slice(startInt until endInt)

        val value = when (slice.size) {
            1 -> slice[0].toUByte().toUInt()
            2 -> {
                val bytes = slice.toByteArray()
                ((bytes[0].toUByte().toUInt()) or
                    (bytes[1].toUByte().toUInt() shl 8))
            }

            3 -> {
                val bytes = slice.toByteArray()
                ((bytes[0].toUByte().toUInt()) or
                    (bytes[1].toUByte().toUInt() shl 8) or
                    (bytes[2].toUByte().toUInt() shl 16))
            }

            4 -> {
                val bytes = slice.toByteArray()
                ((bytes[0].toUByte().toUInt()) or
                    (bytes[1].toUByte().toUInt() shl 8) or
                    (bytes[2].toUByte().toUInt() shl 16) or
                    (bytes[3].toUByte().toUInt() shl 24))
            }

            else -> throw IllegalStateException("Invalid entry size")
        }

        return ProgramCounter(value)
    }

    /**
     * Returns an iterator over the jump table entries
     */
    override fun iterator(): Iterator<ProgramCounter> = object : Iterator<ProgramCounter> {
        private var index: UInt = 0u

        override fun hasNext(): Boolean =
            index < len()

        override fun next(): ProgramCounter {
            val value = getByIndex(index) ?: throw NoSuchElementException("No more elements")
            index++
            return value
        }
    }
}

File: modules/pvm/src/main/scala/io/forge/jam/pvm/program/JumpTable.scala
package io.forge.jam.pvm.program

import io.forge.jam.pvm.Abi

/**
 * Jump table for PVM programs.
 *
 * The jump table maps addresses to program counter values for indirect jumps.
 * Each entry is a fixed-size little-endian integer (1-4 bytes).
 *
 * @param blob The underlying byte array containing jump addresses
 * @param entrySize Size of each jump table entry in bytes (1-4)
 */
final case class JumpTable(blob: Array[Byte], entrySize: Int) extends Iterable[Int]:

  /**
   * Check if the jump table is empty.
   */
  override def isEmpty: Boolean = len == 0

  /**
   * Get the number of entries in the jump table.
   */
  def len: Int =
    if entrySize == 0 then 0
    else blob.length / entrySize

  /**
   * Get a program counter by memory address.
   *
   * Converts a memory address to a jump table index and returns the target.
   *
   * @param address The memory address (must be aligned)
   * @return Some(target) if valid, None otherwise
   */
  def getByAddress(address: Int): Option[Int] =
    if (address & (Abi.VmCodeAddressAlignment - 1)) != 0 || address == 0 then None
    else getByIndex((address - Abi.VmCodeAddressAlignment) / Abi.VmCodeAddressAlignment)

  /**
   * Get a program counter by index.
   *
   * @param index The entry index
   * @return Some(target) if valid, None otherwise
   */
  def getByIndex(index: Int): Option[Int] =
    if entrySize == 0 then None
    else
      val start = index.toLong * entrySize
      if start > Int.MaxValue then None
      else
        val end = start + entrySize
        if end > Int.MaxValue then None
        else
          val startInt = start.toInt
          val endInt = end.toInt

          if startInt >= blob.length || endInt > blob.length then None
          else
            val value = entrySize match
              case 1 => blob(startInt) & 0xFF
              case 2 =>
                (blob(startInt) & 0xFF) |
                  ((blob(startInt + 1) & 0xFF) << 8)
              case 3 =>
                (blob(startInt) & 0xFF) |
                  ((blob(startInt + 1) & 0xFF) << 8) |
                  ((blob(startInt + 2) & 0xFF) << 16)
              case 4 =>
                (blob(startInt) & 0xFF) |
                  ((blob(startInt + 1) & 0xFF) << 8) |
                  ((blob(startInt + 2) & 0xFF) << 16) |
                  ((blob(startInt + 3) & 0xFF) << 24)
              case _ => throw new IllegalStateException(s"Invalid entry size: $entrySize")
            Some(value)

  /**
   * Iterator over all jump table entries.
   */
  override def iterator: Iterator[Int] = new Iterator[Int]:
    private var index = 0

    override def hasNext: Boolean = index < len

    override def next(): Int =
      val value = getByIndex(index).getOrElse(
        throw new NoSuchElementException("No more elements")
      )
      index += 1
      value

object JumpTable:
  /** Empty jump table */
  val Empty: JumpTable = JumpTable(Array.emptyByteArray, 0)

  /**
   * Create a jump table from a byte array and entry size.
   *
   * @param blob The byte array
   * @param entrySize Entry size in bytes
   * @return JumpTable instance
   */
  def apply(blob: Array[Byte], entrySize: Int): JumpTable =
    new JumpTable(blob, entrySize)
